pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go

```diff
@@ -84,6 +118,118 @@ func createRawDeployment(componentMeta metav1.ObjectMeta,
 	return deployment
 }
 
+func GetKServeContainerPort(podSpec *corev1.PodSpec) string {
+	for _, container := range podSpec.Containers {
+		if container.Name == "kserve-container" {
+			if len(container.Ports) > 0 {
+				return strconv.Itoa(int(container.Ports[0].ContainerPort))
+			}
+		}
+	}
+	return ""
+}
+
+func generateOauthProxyContainer(clientset kubernetes.Interface, isvc string, namespace string, upstreamPort string) (corev1.Container, error) {
+	oauthImage := constants.OauthProxyImage
+	oauthCpuLimit := constants.OauthProxyResourceCPULimit
+	oauthMemoryLimit := constants.OauthProxyResourceMemoryLimit
+	oauthCpuRequest := constants.OauthProxyResourceCPURequest
+	oauthMemoryRequest := constants.OauthProxyResourceMemoryRequest
+	inferenceServiceConfigMap, err := clientset.CoreV1().ConfigMaps(constants.KServeNamespace).Get(context.TODO(), constants.InferenceServiceConfigMapName, metav1.GetOptions{})
+	configDataSuccess := false
+	if err == nil {
+		var oauthData map[string]interface{}
+		if err := json.Unmarshal([]byte(inferenceServiceConfigMap.Data["deploy"]), &oauthData); err == nil {
+			configDataSuccess = true
+		}
+		if configDataSuccess {
+			if str, ok := oauthData["image"].(string); ok {
+				oauthImage = str
+			}
+			if str, ok := oauthData["cpuLimit"].(string); ok {
+				oauthCpuLimit = str
+			}
+			if str, ok := oauthData["memoryLimit"].(string); ok {
+				oauthMemoryLimit = str
+			}
+			if str, ok := oauthData["cpuRequest"].(string); ok {
+				oauthCpuRequest = str
+			}
+			if str, ok := oauthData["memoryRequest"].(string); ok {
+				oauthMemoryRequest = str
+			}
+		}
+	}
+
+	return corev1.Container{
+		Name: "oauth-proxy",
+		Args: []string{
+			`--https-address=:8443`,
+			`--provider=openshift`,
+			`--skip-provider-button`,
+			`--openshift-service-account=kserve-sa`,
+			`--upstream=http://localhost:` + upstreamPort,
```

Comment:
Looks like the upstream is always configured to the predictor container (the runtime). But when observing here: https://github.com/opendatahub-io/kserve/blob/4add27b4fcdc34a2ae4ceb8e50d6838db6ec8651/pkg/controller/v1beta1/inferenceservice/reconcilers/service/service_reconciler.go#L112-L123

I can see that the `Service` is choosing between the agent container and the predictor container. Since when auth is enabled, that `Service` will be moved to oauth-proxy, there should be an equivalent logic here, so that oauth-proxy chooses between the predictor and the agent container.

By setting it fixed to the predictor, the agent container can be bypassed, leading to batching and logging features to stop working when enabling auth. The logging feature is being used by TrustyAI team, so I think we should ensure logging still works.